LSP/references: Add context to locations returned by server

mfussenegger opened the pull request:
This changes the `textDocument/references` callback to annotate the
locations returned by the server with the content present at the
locations range.

The quickfix list then looks as follows:

    cr8/insert_fake_data.py|17 col 32| from .misc import parse_table, parse_version
    cr8/insert_fake_data.py|43 col 15| version = parse_version(r['rows'][0][0])
    cr8/java_magic.py|8 col 22| from cr8.misc import parse_version
    cr8/java_magic.py|30 col 19| version = parse_version(fst)
    cr8/java_magic.py|33 col 16| return parse_version(version_str)

Instead of:

    cr8/insert_fake_data.py|17 col 32|
    cr8/insert_fake_data.py|43 col 15|
    cr8/java_magic.py|8 col 22|
    cr8/java_magic.py|30 col 19|
    cr8/java_magic.py|33 col 16|

mfussenegger committed:
LSP/references: Add context to locations returned by server

This changes the `textDocument/references' callback to annotate the
locations returned by the server with the content present at the
locations range.

The quickfix list then looks as follows:

    cr8/insert_fake_data.py|17 col 32| from .misc import parse_table, parse_version
    cr8/insert_fake_data.py|43 col 15| version = parse_version(r['rows'][0][0])
    cr8/java_magic.py|8 col 22| from cr8.misc import parse_version
    cr8/java_magic.py|30 col 19| version = parse_version(fst)
    cr8/java_magic.py|33 col 16| return parse_version(version_str)

Instead of:

    cr8/insert_fake_data.py|17 col 32|
    cr8/insert_fake_data.py|43 col 15|
    cr8/java_magic.py|8 col 22|
    cr8/java_magic.py|30 col 19|
    cr8/java_magic.py|33 col 16|

mfussenegger commented on runtime/lua/vim/lsp/callbacks.lua:
```diff
@@ -37,6 +37,19 @@ end
 
 M['textDocument/references'] = function(_, _, result)
   if not result then return end
+  for _, loc in ipairs(result) do
```
Wasn't sure if this could even be integrated in https://github.com/neovim/neovim/blob/ab12a229410c74e3bcc469bd34a3a4c40af8857c/runtime/lua/vim/lsp/util.lua#L736 there it would also affect other operations.

h-michael replied:
I think we should move this operation to `M.locations_to_items`.
`M.locations_to_items` is used by only `M.set_loclist` and `M.set_qflist`.
I think this operation is useful for both.

mfussenegger replied:
I've pushed another commit that integrates it into `M.locations_to_items` - made the change also a lot simpler because there `line` is already accessible.

h-michael commented on runtime/lua/vim/lsp/callbacks.lua:
```diff
@@ -37,6 +37,19 @@ end
+    if loc.uri and not loc.message then
+      local bufnr = vim.uri_to_bufnr(loc.uri)
+      if not api.nvim_buf_is_loaded(bufnr) then
+        vim.fn.bufload(bufnr)
```
I think if load unloaded buffers, should `bdelete` those buffers after get line.

mfussenegger commented on runtime/lua/vim/lsp/callbacks.lua:
```diff
@@ -37,6 +37,19 @@ end
 M['textDocument/references'] = function(_, _, result)
   if not result then return end
+  for _, loc in ipairs(result) do
+    if loc.uri and not loc.message then
```
I think I had anticipated to use this in `M.locations_to_items` which people can also use for diagnostics, where a message could already be present.

h-michael commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -772,7 +772,7 @@ function M.locations_to_items(locations)
             lnum = row + 1,
             col = col + 1;
-            text = temp.msg;
+            text = temp.msg or line;
```
I think both Location and LocationLink don't have message field.
So can we delete codes about msg?

mfussenegger replied:
I think the `message` field is related to having used `set_loclist` to populate diagnostics: 

https://github.com/neovim/neovim/blob/ff1730373c6139db14b8f2f9b24d4ccd7fcfb01d/runtime/lua/vim/lsp/callbacks.lua#L35

I also use an override like:

```lua
local default_diagnostics_callback = lsp.callbacks['textDocument/publishDiagnostics']
local function diagnostics_callback(err, method, result, client_id)
    default_diagnostics_callback(err, method, result, client_id)
    if result and result.diagnostics and result.uri then
        local current_buf = api.nvim_get_current_buf()
        if vim.uri_to_bufnr(result.uri) == current_buf and myutil.exists(vim.uri_to_fname(result.uri)) then
            for _, v in ipairs(result.diagnostics) do
                v.uri = v.uri or result.uri
            end
            util.set_loclist(result.diagnostics)
        end
    end
end
```

But I agree that a method called `locations_to_items` should not deal with turning diagnostics into items as well. But maybe removing the message can be a follow up in a separate PR?

It probably makes sense to provide some kind of alternative for people who want to have the diagnostics in the location list.

h-michael replied:
I understand your thought :)
Thank you for the detailed explanation.

mfussenegger replied:
I'll follow up on this. I already toyed around a bit to get an idea how the current diagnostics functions could be changed to make them more customizable & composable.  (https://github.com/mfussenegger/dotfiles/compare/lsp-diag?expand=1#diff-e9e37b70bcd90d6df21d11736e98c583R19) 

My thoughts right now go into the direction of:

- Having the default publishDiagnostics callback as is
- Provide access to the diagnostics (bufnr -> linenr -> list of diagnostics)
- Make the logic that is currently invoked by the publishDiagnostics more composable, so users can replace the `publishDiagnostics` callback and easily re-use the parts they'd like to keep (e.g, store the diagnostics, active highlights)

mfussenegger replied:
Rebased and added a second commit to drop the `message` handling.

I'll create a PR that adds something like 

```
local function diagnostics_to_items(bufnr, buf_diagnostics)
    local items = {}
    for linenr, diagnostics in pairs(buf_diagnostics) do
        if #diagnostics > 0 then
            local d = diagnostics[1]
            table.insert(items, {
                bufnr = bufnr,
                lnum = linenr + 1,
                vcol = 1,
                col = d.range.start.character + 1,
                text = d.message
            })
        end
    end
    return items
end
```

depending on the outcome of https://github.com/neovim/neovim/pull/11932

h-michael commented:
@justinmk lgtm

mfussenegger committed:
LSP: Remove diagnostic message handling in locations_to_items

`locations_to_items` is for turning `Location[]` into items, not for
`Diagnostic[]`

teto commented:
rebase please, I will have a look at it

mfussenegger commented:
@teto Just rebased. Thank you for having a look

teto commented:
tested and works. I can't restart the jobs for some reason but the fialures look unrelated. Thanks for the PR.

teto commented:
I didn't realize but the change removed printing of the diagnostic in quickfix. Thus this fixes one usecase but breaks the otherone. There should be a way to have both and use either `diagnostic` or `line` as the text.

mfussenegger commented:
> I didn't realize but the change removed printing of the diagnostic in quickfix. Thus this fixes one usecase but breaks the other ones. There should be a way to have both and use either `diagnostic` or `line` as the text.
> 
> For now I've reverted locally [3820165](https://github.com/neovim/neovim/commit/38201650cd60c001c4b639f30ca45f1b8eba98ca) because I would like to add tests on top of the fix.


@teto Yes, this was discussed here: https://github.com/neovim/neovim/pull/11910#discussion_r382539781


I wanted to follow up on this with something like this:

```lua
local function diagnostics_to_items(bufnr, buf_diagnostics)
    local items = {}
    if not buf_diagnostics then return items end
    for linenr, diagnostics in pairs(buf_diagnostics) do
        if #diagnostics > 0 then
            local d = diagnostics[1]
            table.insert(items, {
                bufnr = bufnr,
                lnum = linenr + 1,
                vcol = 1,
                col = d.range.start.character + 1,
                text = d.message
            })
        end
    end
    return items
end
```

But wanted to do that on top of https://github.com/neovim/neovim/pull/11932  (also with tests added)
